crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,387 @@
+            None => self.pipeline_run_needed(),
+        };
+        Ok(())
+    }
+}
+
+/// On initialization, the controller will poll the message receiver and return [Poll::Pending]
+/// until the first forkchoice update message is received.
+///
+/// As soon as the controller receives the first forkchoice updated message and updates the local
+/// forkchoice state, it will launch the pipeline to sync to the head hash.
+/// While the pipeline is syncing, the controller will keep processing messages from the receiver
+/// and forwarding them to the blockchain tree.
+impl<DB, U, C, EF> Future for BeaconConsensusEngine<DB, U, C, EF>
+where
+    DB: Database + Unpin + 'static,
+    U: SyncStateUpdater + Unpin + 'static,
+    C: Consensus + Unpin,
+    EF: ExecutorFactory + Unpin + 'static,
+{
+    type Output = Result<(), BeaconEngineError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        // Process all incoming messages first.
+        while let Poll::Ready(Some(msg)) = this.message_rx.poll_next_unpin(cx) {
+            match msg {
+                BeaconEngineMessage::ForkchoiceUpdated(state, tx) => {
+                    let response = this.on_forkchoice_updated(state);
+                    let _ = tx.send(response);
+                }
+                BeaconEngineMessage::NewPayload(block, tx) => {
+                    let response = this.on_new_payload(block);
+                    let _ = tx.send(response);
+                }
+            }
+        }
+
+        // Set the next pipeline state.
+        loop {
+            // Lookup the forkchoice state. We can't launch the pipeline without the tip.
+            let forckchoice_state = match &this.forkchoice_state {
+                Some(state) => state,
+                None => return Poll::Pending,
+            };
+
+            let tip = forckchoice_state.head_block_hash;
+            let next_state = match this.pipeline_state.take().expect("pipeline state is set") {
+                PipelineState::Running(mut fut) => {
+                    match fut.poll_unpin(cx) {
+                        Poll::Ready((pipeline, result)) => {
+                            // Any pipeline error at this point is fatal.
+                            if let Err(error) = result {
+                                return Poll::Ready(Err(error.into()))
+                            }
+
+                            // Update the state and hashes of the blockchain tree if possible
+                            if let Err(error) = this
+                                .restore_tree_if_possible(forckchoice_state.finalized_block_hash)
+                            {
+                                return Poll::Ready(Err(error.into()))
+                            }
+
+                            // Get next pipeline state.
+                            this.next_pipeline_state(pipeline, tip)
```

Comment:
If the current pipeline tip is the same as `tip` should this set the state to `Idle`?